pkg/clients/ec2/securitygroup.go

```diff
@@ -129,28 +175,107 @@ func LateInitializeSG(in *v1beta1.SecurityGroupParameters, sg *ec2.SecurityGroup
 	in.GroupName = awsclients.LateInitializeString(in.GroupName, sg.GroupName)
 	in.VPCID = awsclients.LateInitializeStringPtr(in.VPCID, sg.VpcId)
 
-	if len(in.Egress) == 0 && len(sg.IpPermissionsEgress) != 0 {
-		in.Egress = v1beta1.BuildIPPermissions(sg.IpPermissionsEgress)
+	// If there is a mismatch in lengths, then it's a sign for desire change
+	// that should be handled via add or remove.
+	if len(in.Egress) == len(sg.IpPermissionsEgress) {
+		in.Egress = LateInitializeIPPermissions(in.Egress, sg.IpPermissionsEgress)
 	}
-
-	if len(in.Ingress) == 0 && len(sg.IpPermissions) != 0 {
-		in.Ingress = v1beta1.BuildIPPermissions(sg.IpPermissions)
+	if len(in.Ingress) == len(sg.IpPermissions) {
+		in.Ingress = LateInitializeIPPermissions(in.Ingress, sg.IpPermissions)
 	}
 
 	if len(in.Tags) == 0 && len(sg.Tags) != 0 {
 		in.Tags = v1beta1.BuildFromEC2Tags(sg.Tags)
 	}
 }
 
+// LateInitializeIPPermissions returns an array of []v1beta1.IPPermission whose
+// empty optional fields are filled with what's observed in []ec2.IpPermission.
+//
+// Note that since there is no unique identifier for each IPPermission, its order
+// is assumed to be stable and used indexes are used as identifier.
```

Comment:
nitpick:
```suggestion
// is assumed to be stable and indexes are used as identifier.
```

Reply:
Any idea if AWS sorts rules in a consistent order? Seems like if we don't mirror then it will look like we always need update.

Reply:
No one knows, really. This is the problem with resources without any identifiers. Though manually sorting both of the arrays could be a good idea, it will be meaningless when we separate those rules into their [own types](https://github.com/crossplane/provider-aws/issues/297). Let me see if I can write something to address this.

Reply:
Fair point, I just don't want to give the illusion to users that we support this inline for now if in reality it is going to thrash because that could cause issues with other requests.

Reply:
In regards to sorting, it seems like we should establish uniqueness of an entry on the fromPort/toPort/IpProtocol combination.

Reply:
> it seems like we should establish uniqueness of an entry on the fromPort/toPort/IpProtocol combination.

I'm not sure. Looking at comments of those fields, even `IpProtocol` can get `-1` value so they don't really match their actual configurations. For example, I've tested that there could be a case where the port range is `0 - 3306`, protocol `TCP` but one referring to a security group as source and another one referring to a IpV4 CIDR block as source.

Reply:
Alright, I've tried a few things to sort correctly and it gets really ugly. I feel like I spent way more time than necessary on this now. I think it's OK to not sort it and rely on AWS and this is still improvement over the existing behavior.

Reply:
Fair enough. I guess worst case scenario we only send a successful update every so often. Feel free to merge when ready then I can start release proceedings :+1: